### PR TITLE
Checkpointer revival

### LINKED ISSUE #324: Barebones checkpointing
We currently have no checkpointing anymore. We need a barebones checkpointer that only checkpoints and saves the essentials. Should be more transparent than serializing the `Model` to disk, which also comes with multiple issues.

Should also be useful for long runs (e.g. seasonal cycle simulations) and more rigorous regression tests that actually test the dynamics.

### PR DESCRIPTION
**Pull request by ali-ramadhan**
This PR reintroduces checkpointing. This time the checkpointer saves the bare essentials to a JLD2 file.

Checkpointer will not save forcing functions as before (see #141).

Need to test GPU checkpointing, checkpointing with LES closures (I assumed `model.diffusivities` does not need to be checkpointed), and checkpointing with fancier boundary conditions.

A major issue with this checkpointer is that it can't restore the largest models because it creates a `Model` and then copies the data from the checkpoint file into the model fields. But if the model is using up all memory, then there's no room for reading data from disk to memory.

It should feed the data directly through the Model constructor, but this would require refactoring some of the model and field code.

This can be addressed in this PR or in a future PR, although it's kind of useless because checkpointing becomes more important the larger the model/simulation...

Resolves #324

### CODE REVIEW

**Comment by ali-ramadhan** [comment]
I think `output_writers.jl` is getting pretty cluttered (almost 400 lines with functionality shared between output writers) so I might try to move them to individual files in a future PR.

**Comment by glwagner** [comment]
I’m happy with output_writers.jl; I think all the code in that file is appropriately related.

The checkpointer could provide its own constructor to avoid excess memory allocation. 

For GPU problems I don’t think there is an issue: checkpointed arrays can b loaded into CPU memory rather than GPU memory, and then the data can be copied into the fields allocated by the model constructor. So at first glance the excess memory allocation does not seem like a major issue on modern CPUs. 

I am particularly concerned about the maintainability of the checkpointer, since it will need to be updated every time a new feature is added. Let’s make sure the design is optimally maintainable before merging.

File: src/output_writers.jl [hunk 1]
```diff
@@ -324,3 +324,72 @@ function time_to_write(clock, out::JLD2OutputWriter)
...
+end
+
+function Checkpointer(; output_frequency, dir=".", prefix="checkpoint", force=false)
+    mkpath(dir)
+    return Checkpointer(dir, prefix, output_frequency)
+end
+
+function savesubfields!(file, model, name, flds=propertynames(getproperty(model, name)))
+    for f in flds
+        file["$name/$f"] = Array(getproperty(getproperty(model, name), f).data.parent)
```
Comment (glwagner) [inline src/output_writers.jl]: Put an `if` statement that does not save a field if its type is `Function`?

File: src/output_writers.jl [hunk 2]
```diff
@@ -324,3 +324,72 @@ function time_to_write(clock, out::JLD2OutputWriter)
...
     end
 end
+
+mutable struct Checkpointer <: OutputWriter
+                 dir :: String
+              prefix :: String
+    output_frequency :: Int
+end
+
+function Checkpointer(; output_frequency, dir=".", prefix="checkpoint", force=false)
```
Comment (glwagner) [inline src/output_writers.jl]: The kwarg 'force' is not used here. It should either be used, or removed from the function signature.

In the `JLD2OutputWriter`, the kwarg `force` indicates whether file creation should be 'forced' (it corresponds to the same keyword passed to `mkpath`.

File: src/output_writers.jl [hunk 3]
```diff
@@ -324,3 +324,72 @@ function time_to_write(clock, out::JLD2OutputWriter)
...
+        file["$name/$f"] = Array(getproperty(getproperty(model, name), f).data.parent)
+    end
+    return nothing
+end
+
+checkpointed_structs   = [:arch, :boundary_conditions, :grid, :clock, :eos, :constants, :closure]
+checkpointed_fieldsets = [:velocities, :tracers, :G, :Gp]
+
+function write_output(model, c::Checkpointer)
+    @warn "Checkpointer will not save forcing functions, output writers, or diagnostics. They will need to be " *
```
Comment (glwagner) [inline src/output_writers.jl]: Does it make sense to have a warning that is *always* printed? Perhaps is it simply better to document this aspect of the checkpointer?

File: src/output_writers.jl [hunk 4]
```diff
@@ -324,3 +324,72 @@ function time_to_write(clock, out::JLD2OutputWriter)
...
+        # Checkpointing structs containing fields.
+        [savesubfields!(file, model, p) for p in checkpointed_fieldsets]
+    end
+end
+
+_arr(::CPU, a) = a
+_arr(::GPU, a) = CuArray(a)
+
+function restore_from_checkpoint(filepath)
+    @warn "Checkpointer cannot restore forcing functions, output writers, or diagnostics. They will need to be " *
```
Comment (glwagner) [inline src/output_writers.jl]: See above. What we need is documentation about proper checkpointing, which will include information about how to restore a model from a checkpoint that includes functions. In fact, we could even provide features in the checkpointer that streamline this process (by indicating parts of the model structure that are associated with functions, and asking the user to provide those functions during checkpoint restoration).

File: test/test_output_writers.jl [hunk 5]
```diff
@@ -35,11 +35,65 @@ function run_thermal_bubble_netcdf_tests()
...
+    true_model = Model(N=(Nx, Ny, Nz), L=(Lx, Ly, Lz), ν=4e-2, κ=4e-2, arch=arch)
+
+    # Add a cube-shaped warm temperature anomaly that takes up the middle 50%
+    # of the domain volume.
+    i1, i2 = round(Int, Nx/4), round(Int, 3Nx/4)
+    j1, j2 = round(Int, Ny/4), round(Int, 3Ny/4)
+    k1, k2 = round(Int, Nz/4), round(Int, 3Nz/4)
+    true_model.tracers.T.data[i1:i2, j1:j2, k1:k2] .+= 0.01
+
+    checkpointed_model = deepcopy(true_model)
```
Comment (glwagner) [inline test/test_output_writers.jl]: Why is it necessary to copy the model before checkpointing?

File: src/output_writers.jl [hunk 6]
```diff
@@ -324,3 +324,72 @@ function time_to_write(clock, out::JLD2OutputWriter)
...
+
+function savesubfields!(file, model, name, flds=propertynames(getproperty(model, name)))
+    for f in flds
+        file["$name/$f"] = Array(getproperty(getproperty(model, name), f).data.parent)
+    end
+    return nothing
+end
+
+checkpointed_structs   = [:arch, :boundary_conditions, :grid, :clock, :eos, :constants, :closure]
+checkpointed_fieldsets = [:velocities, :tracers, :G, :Gp]
```
Comment (glwagner) [inline src/output_writers.jl]: `G` and `Gp` are not fields of `model`.

File: src/output_writers.jl [hunk 7]
```diff
@@ -324,3 +324,72 @@ function time_to_write(clock, out::JLD2OutputWriter)
...
+
+        # Checkpointing structs containing fields.
+        [savesubfields!(file, model, p) for p in checkpointed_fieldsets]
+    end
+end
+
+_arr(::CPU, a) = a
+_arr(::GPU, a) = CuArray(a)
+
+function restore_from_checkpoint(filepath)
```
Comment (glwagner) [inline src/output_writers.jl]: This function must have a `; kwargs...` argument which is then merged with the retrieved `kwargs` from the checkpoint file before being passed to the model constructor. This is needed to restore models with forcing functions or non-default boundary conditions.

File: src/output_writers.jl [hunk 8]
```diff
@@ -324,3 +324,72 @@ function time_to_write(clock, out::JLD2OutputWriter)
...
+    @warn "Checkpointer will not save forcing functions, output writers, or diagnostics. They will need to be " *
+          "restored manually."
+
+    filepath = joinpath(c.dir, c.prefix * string(model.clock.iteration) * ".jld2")
+
+    jldopen(filepath, "w") do file
+        # Checkpointing model properties that we can just serialize.
+        [file["$p"] = getproperty(model, p) for p in checkpointed_structs]
+
+        # Checkpointing structs containing fields.
```
Comment (glwagner) [inline src/output_writers.jl]: Perhaps this step can be combined with the first, along with an if statement that inspects the content of the struct to be saved and determines whether it contains `Fields` or not.

File: src/output_writers.jl [hunk 9]
```diff
@@ -324,3 +324,72 @@ function time_to_write(clock, out::JLD2OutputWriter)
...
+end
+
+function savesubfields!(file, model, name, flds=propertynames(getproperty(model, name)))
+    for f in flds
+        file["$name/$f"] = Array(getproperty(getproperty(model, name), f).data.parent)
+    end
+    return nothing
+end
+
+checkpointed_structs   = [:arch, :boundary_conditions, :grid, :clock, :eos, :constants, :closure]
```
Comment (glwagner) [inline src/output_writers.jl]: These fields could be made properties of the `Checkpointer`. This will allow the user to add/remove fields from checkpointing, which may be generally useful in the future and is currently useful for omitting boundary conditions from checkpointing.

File: src/output_writers.jl (hunk 1 above)
Comment (ali-ramadhan) [reply src/output_writers.jl]: Sounds like a good check. Won't be needed as long we maintain `checkpointed_fieldsets` but good check either way.

File: src/output_writers.jl (hunk 3 above)
Comment (ali-ramadhan) [reply src/output_writers.jl]: Hmmm, might then be good to just print the warning if there is a non-zero forcing function, or an output writer or diagnostic included.

Then yeah warning may be removed if the checkpointer is well-documented.

File: src/output_writers.jl (hunk 4 above)
Comment (ali-ramadhan) [reply src/output_writers.jl]: Yeah warning doesn't have be in `restore_from_checkpoint`. Passing forcing functions is a good idea.

File: test/test_output_writers.jl (hunk 5 above)
Comment (ali-ramadhan) [reply test/test_output_writers.jl]: It's creating a second model that will be checkpointed as opposed to `true_model` which isn't.

Probably over paranoid but I wanted the two models to be time-stepped separately.

File: src/output_writers.jl (hunk 6 above)
Comment (ali-ramadhan) [reply src/output_writers.jl]: True. This was before PR #325 was merged and included the tendencies with `AdamsBashforthTimestepper`.

**Comment by ali-ramadhan** [comment]
I'll work on this until we're happy with the interace but will leave restoring massive models (that fill up memory) properly for another PR. 

> I’m happy with output_writers.jl; I think all the code in that file is appropriately related.

I still find it a little messy and hard to navigate. Maybe a good solution for now would be to reorganize the file into sections, and shared functions can be moved to the top.

> I am particularly concerned about the maintainability of the checkpointer, since it will need to be updated every time a new feature is added. Let’s make sure the design is easy to maintain before merging.

Eventually the `Model` struct will become stable but until then we'll have to modify the checkpointer to account for that. But yeah, having something flexible and easy to modify is important.

> add a field (perhaps `checkpointed_fields`) to the `Checkpointer` that allows the user to control which subfields of `model` are checkpointed.

Might make more sense. In general, the checkpointer should be flexible but you'd probably only need to change what's checkpointed for esoteric cases. E.g. if you remove something important, you'll

File: src/output_writers.jl (hunk 1 above)
Comment (glwagner) [reply src/output_writers.jl]: I think that we should allow users to supply the `modelfields` that are to be checkpointed. In that case, such a check will be important.

File: src/output_writers.jl (hunk 3 above)
Comment (glwagner) [reply src/output_writers.jl]: Here is what I propose:

1. Allow users to set the `modelfields` that are to be checkpoints as an argument to the `Checkpointer` constructor

2. If one of those fields that the user has asked to be saved contains a function, emit a warning.

File: src/output_writers.jl (hunk 1 above)
Comment (glwagner) [reply src/output_writers.jl]: Continuing from my comment below, the `if`-statement can also emit a warning that field `x` will not be saved.

File: test/test_output_writers.jl (hunk 5 above)
Comment (glwagner) [reply test/test_output_writers.jl]: Are you worried that checkpointing will modify `model`?

Since this behavior is undesirable / unexpected, perhaps it should simply be tested for, so that we can assume that the model is not modified.

**Comment by glwagner** [comment]
> Eventually the Model struct will become stable but until then we'll have to modify the checkpointer to account for that.

This is not the only issue. Another issue is if new subfields of model are added that contain non-checkpointable elements, like forcing functions. In that sense the `Model` struct will never be "stable" with respect to the types of data it references. For example, there could someday be a `grid` that contains forcing functions. Also, we should checkpoint boundary conditions by default. However, because boundary conditions can contain forcing functions, we need to be able to robustly/programmatically save only those boundary conditions that do not contain references to functions.

To address this, we need a generalized checkpointer that works in a wide range of circumstances, is relatively automated, modular, and can be user-modified.

I'm happy to help develop these features. One key is a function that can query a struct recursively to determine if it contains any references to `Function`s at any level. A second element is control to save an entire object in the case that the object contains no references to functions, versus a different but useful secondary behavior if the object *does* contain a reference to functions.

**Comment by ali-ramadhan** [comment]
I've changed things a bit, Mainly, the structs and fieldsets to checkpoint can be specified through the `Checkpointer` constructor. If an incompatible struct is provided, e.g. `:forcing`, an exception is raised. Validation only occurs in the constructor so no need for if-statements and warnings anywhere else.

@glwagner Lemme know if this looks okay to merge.

File: src/output_writers.jl (hunk 1 above)
Comment (glwagner) [reply src/output_writers.jl]: @ali-ramadhan was this resolved?

File: src/output_writers.jl (hunk 3 above)
Comment (glwagner) [reply src/output_writers.jl]: @ali-ramadhan was this resolved?

File: src/output_writers.jl (hunk 6 above)
Comment (glwagner) [reply src/output_writers.jl]: @ali-ramadhan was this resolved?

File: src/output_writers.jl (hunk 9 above)
Comment (glwagner) [reply src/output_writers.jl]: @ali-ramadhan was this resolved?

File: src/output_writers.jl [hunk 10]
```diff
@@ -392,19 +392,29 @@ VerticalPlanes(model) = VerticalPlanes(model.arch, model.grid)
...
+                      checkpointed_fieldsets = (:velocities, :tracers, :G, :Gp),
+                      force=false)
     mkpath(dir)
     return Checkpointer(dir, prefix, output_frequency)
 end
 
 function savesubfields!(file, model, name, flds=propertynames(getproperty(model, name)))
-    for f in flds
-        file["$name/$f"] = Array(getproperty(getproperty(model, name), f).data.parent)
+    if name ∉ (:forcing)
```
Comment (glwagner) [inline src/output_writers.jl]: I'm a bit confused about how this function works:

- why restrict this functionality to `Field` (as it looks to me), but provide a hook when `name=:forcing`?
- can't we check the `type` of the object being saved? Other fields (like boundary conditions, or future fields that we add to model) may be functions.

I think it'd be better to write a function that dispatches on the type of the object being saved, eg

```julia
savefield(file, location, field) = file[location] = field
savefield(file, location, field::AbstractArray) = file[location] = Array(field)
savefield(file, location, field::Field) = file[location] = Array(field.data.parent)
savefield(file, location, field::Function) = warn("Cannot checkpoint Function object into $location!")
```

This seems more robust and general to me.
